### PR TITLE
Support per query filter

### PR DESCRIPTION
**Pull request by yashpatel007**
- [ ] Does this PR have a descriptive title that could go in our release notes?
- [ ] Does this PR add any new dependencies?
- [ ] Does this PR modify any existing APIs?
   - [ ] Is the change to the API backwards compatible?
- [ ] Should this result in any changes to our documentation, either updating existing docs or adding new ones?

### CODE REVIEW

File: include/utils.h [hunk 1]
```diff
@@ -1014,6 +1014,71 @@ template <typename T> inline void normalize(T *arr, size_t dim)
         arr[i] = (T)(arr[i] / sum);
     }
 }
+template <typename T> inline std::vector<T> readFileLinesInVector(const std::string &filename, bool unique = false)
```
Comment (harsha-simhadri) [inline include/utils.h:1017]: could we call this read_file_to_vector_of_strings

Comment (harsha-simhadri) [reply include/utils.h:1017]: Also, do we really need to parse as strings as opposed to labelT (which could be set to uint32, uint64)

File: include/utils.h [hunk 2]
```diff
@@ -1014,6 +1014,71 @@ template <typename T> inline void normalize(T *arr, size_t dim)
...
+                if (std::remove(curr_path_to_clean.c_str()) != 0)
+                    diskann::cout << "Warning: Unable to remove file :" << curr_path_to_clean << std::endl;
+            }
+        }
+    }
+    catch (const std::exception &e)
+    {
+        diskann::cout << "Warning: Unable to clean all artifacts" << std::endl;
+    }
+    diskann::cout << "Cleaned all artifacts" << std::endl;
```
Comment (harsha-simhadri) [inline include/utils.h:1080]: control flow broken.

File: tests/search_disk_index.cpp [hunk 3]
```diff
@@ -51,8 +51,8 @@ int search_disk_index(diskann::Metric &metric, const std::string &index_path_pre
                       const std::string &result_output_prefix, const std::string &query_file, std::string &gt_file,
                       const unsigned num_threads, const unsigned recall_at, const unsigned beamwidth,
                       const unsigned num_nodes_to_cache, const _u32 search_io_limit, const std::vector<unsigned> &Lvec,
-                      const float fail_if_recall_below, const bool use_reorder_data = false,
-                      const std::string &filter_label = "")
+                      const float fail_if_recall_below, const std::vector<std::string> &query_filters,
+                      const bool use_reorder_data = false)
```
Comment (harsha-simhadri) [inline tests/search_disk_index.cpp:55]: Is this backward compatiable when there are no filters?

File: tests/utils/compute_groundtruth.cpp [hunk 4]
```diff
@@ -421,24 +427,104 @@ inline void parse_label_file_into_vec(size_t &line_cnt, const std::string &map_f
 }
 
 template <typename T>
-int aux_main(const std::string &base_file, const std::string &label_file, const std::string &query_file,
-             const std::string &gt_file, size_t k, const std::string &filter_label, const std::string &universal_label,
-             const diskann::Metric &metric, const std::string &tags_file = std::string(""))
+std::vector<std::vector<std::pair<uint32_t, float>>> processUnfilteredParts(const std::string &base_file,
```
Comment (harsha-simhadri) [inline tests/utils/compute_groundtruth.cpp:430]: Can we have two compute GT tools? One for unfiltered and filtered?

File: include/utils.h (hunk 1 above)
Comment (yashpatel007) [reply include/utils.h:1017]: We could do labelT but, I
 was thinking of a more generic function. Let me translate the file reading logic to respect the template type.

File: tests/search_disk_index.cpp (hunk 3 above)
Comment (yashpatel007) [reply tests/search_disk_index.cpp:55]: Yes, the code is indeed backward compatible. The decision point is when query_filters, is empty.

File: tests/utils/compute_groundtruth.cpp (hunk 4 above)
Comment (yashpatel007) [reply tests/utils/compute_groundtruth.cpp:337]: Now we have two GT files 
For Unfiltered : compute_groundtruth.cpp
for Filtered : compute_groundtruth_for_filtered.cpp

**Review by harsha-simhadri — APPROVED** [review]

**Pull request merged**